terraform/node_resource_plan.go

```diff
@@ -77,6 +78,12 @@ func (n *NodePlannableResource) DynamicExpand(ctx EvalContext) (*Graph, error) {
 		return nil, diags.Err()
 	}
 
+	forEachMap, forEachDiags := evaluateResourceForEachExpression(n.Config.ForEach, ctx)
+	if forEachDiags.HasErrors() {
+		fmt.Println("hit the hasErrors block")
```

Comment:
This remains a question for me -- I added a bit that checks this earlier (at eval_validate), and thus diags get exposed to the user, but I still am confused as to why these diags don't get exposed to the user. Having it does prevent a panic though.

Reply:
Hmm, that _is_ interesting... I think that suggests a bug in how the graph walk deals with `DynamicExpand` where it's not handling returned diagnostics properly, and we've probably not been seeing it just because other codepaths happen to catch all the same errors.

However, I went to find the caller of this and found that at least _that_ call seems to be dealing with it okay:

https://github.com/hashicorp/terraform/blob/86e6481cc661b2ba694d353aab6d11838c066c7c/terraform/graph.go#L103-L107

So if this is a bug then it must be somewhere further up the call stack, I guess.

Reply:
Ok, that's sounding like "not something to deal with here" --- thanks for the additional context, I'm curious when we'll see this show up again 🤔 Not sure how to word an issue to open, but maybe we'll remember this in the future 🤷‍♀